# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.3.1] - 2025-05-26
- Re-deployment to enable labels for optional outputs
- [GRD-948](https://jira.oicr.on.ca/browse/GRD-948)

## [2.3.0] - 2024-06-25
### Added
- [GRD-797](https://jira.oicr.on.ca/browse/GRD-797)] - add vidarr labels to outputs (changes to medata only)

## [2.2.1] - 2023-05-24
### Added
- add optional parameter numReads
- use python module in countChunkSize

## [2.2.0] - 2023-03-01
### Changed
- move version 2.1.2 to 2.2.0

## [2.1.2] - 2023-02-01
### Added
- upgrade bwa version to 0.7.17, re-organized the reference and modules parameters. It is identical to the version 2.1.1 other than the modules.

## [2.1.1] - 2023-02-01
### Changed
- re-organized the reference and modules parameters, bwa version 0.7.12.

## [2.1.0] - 2022-10-13
### Added
- add options for UMI adapter trimming, UMI extract; add a "reference" parameter; add default values for bwaMem_modules and bwaMem_ref based on reference

## [1.0.0] - 2021-05-31
### Changed
- migrate Niassa 2.0.1 to Vidarr 1.0.0

## [2.0.1] - 2020-07-21
### Added
- [GRD-22](https://jira.oicr.on.ca/browse/GRD-22)] - support single-ended reads

## [2.0.0] - 2020-02-24
### Changed
- [GP-2221](https://jira.oicr.on.ca/browse/GP-2221)] - coverting to cromwell wdl, adding an option of splitting fastq files into chunks to speed up the alignment

## [1.2.0] - 2018-09-19
### Added
- [GP-1143](https://jira.oicr.on.ca/browse/GP-1143)] - Provision out the statistics data generated by cutadapt
